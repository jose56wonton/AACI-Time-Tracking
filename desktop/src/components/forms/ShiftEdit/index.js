import ShiftEdit from './shiftEdit';
export default ShiftEdit;