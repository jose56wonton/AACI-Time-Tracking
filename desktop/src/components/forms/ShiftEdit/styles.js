const styles = theme => ({
  gridContainer: {
    width: '100% !important',
    margin: 0
  },
  row: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between'
  },
  headerRow: {
    minHeight: '72px'
  },
  field:{
    margin: 'auto 8px'
  },
  switchBox: {
    width: '100%',
    display: 'flex',
    justifyContent: 'space-around'
  },
  error: { marginTop: '16px' },
  button: {
    marginLeft: '8px'
  },

  ///////////////////////
  textField: { width: '100%' },
  heroContent: {
    width: '100%',
    maxWidth: '1000px'
  },
  error: {
    color: theme.palette.error.main
  },
  spaceBetween: {
    justifyContent: 'space-between'
  },
  flexEnd: {
    alignItems: 'flex-end',
    justifyContent: 'flex-end'
  },
  button: { marginLeft: '10px' },

  verticalCenter: {
    display: 'flex',
    alignItems: 'center'
  },

  formHeader: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginRight: theme.spacing.unit
  },
  formHeaderIcon: {
    fontSize: '32px',
    margin: '0 10px'
  },

  formBody: {
    display: 'flex',
    padding: '12px 0',
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  formElement: {
    marginRight: theme.spacing.unit
  },

  formFooter: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    height:'68px'
  },
  lunchBox: {
    width: '150px'
  },
  fullWidth: {
    width: '100%'
  }
});

export default styles;
