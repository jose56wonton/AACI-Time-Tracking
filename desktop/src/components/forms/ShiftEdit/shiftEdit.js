import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Grid, Typography, Button, Tooltip, IconButton, MenuItem } from '@material-ui/core';
import { Close } from '@material-ui/icons';
import cx from 'classnames';
import { Field, Form, FieldArray } from 'formik';
import { withStyles } from '@material-ui/core/styles';
import { Delete } from '@material-ui/icons';

import TextField from 'components/inputs/TextField';
import Select from 'components/inputs/Select';
import Time from 'components/inputs/Time';

import styles from './styles';

import { minutesToString } from 'helpers/time';

export class ShiftEdit extends Component {
  render() {
    const { classes, label, type, removeShift, isSubmitting, resetForm, initialValues, errors,  values,
      projects, projectTasks, employees,  timeLeft, generalError } = this.props;
    return (
      <Form>
        <Grid container spacing={24} className={classes.gridContainer}>
          <Grid item xs={12} className={cx(classes.headerRow, classes.row)}>
            <Typography variant="h6">{label}</Typography>
            {type === 'edit' && (
              <Tooltip title="Delete">
                <IconButton onClick={removeShift} aria-label="Delete">
                  <Delete />
                </IconButton>
              </Tooltip>
            )}
          </Grid>
          <Grid item xs={12} className={classes.row}>
            <Field
              name='employeeId'
              component={Select}
              items={employees}
              fullWidth
              className={classes.field}
              label="Employee"
            />
            <Field
              name="clockInDate"
              component={TextField}
              margin="none"
              label="Start Date"
              type="datetime-local"
              className={classes.field}
              helper="normal"
            />
            <Field
              name="clockOutDate"
              component={TextField}
              margin="none"
              label="Start Date"
              type="datetime-local"
              defaultValue="2017-05-24T10:30"
              className={classes.field}
              helper="normal"
            />
          </Grid>


          <FieldArray
            name="activities"
            render={arrayHelpers => {
              return (
                <Grid item xs={12} container className={classes.body}>
                  {values.activities &&
                    values.activities.map((activity, index) => {
                      return (
                        <Grid item xs={12}
                          key={index}
                          className={cx(
                            classes.card,
                            classes.verticalCenterBox,
                          )}
                        >
                          <div className={cx(classes.row,classes.bodyRow)}>
                            <Field
                              name={`activities.${index}.projectId`}
                              component={Select}
                              items={projects}
                              fullWidth
                              label="Project"     
                              id={`project-field-${index}`}                                                      
                              className={classes.field}
                              onChange={ () => {                                  
                                arrayHelpers.form.setFieldValue(
                                  `activities.${index}.projectTaskId`,
                                  -1,
                                );                                  
                              }}
                            />
                            <Field
                              name={`activities.${index}.projectTaskId`}
                              component={Select}
                              fullWidth
                              label="Task"
                              className={classes.field}
                            >
                              {
                                projectTasks // This code iterates the projectTask 
                                  .filter(projectTask => {
                                    return activity.projectId === projectTask.projectId; // filters based on project selected
                                  })
                                  .map((projectTask, i) => { // maps those elements
                                    return (
                                      <MenuItem
                                        key={i}
                                        id="projectTaskId"
                                        value={projectTask.id}
                                      >
                                        {projectTask.task.name}
                                      </MenuItem>
                                    );
                                  })
                              }
                            </Field>
                            <Field
                              name={`activities.${index}.length`}
                              component={Time}
                              fullWidth
                              className={classes.field}
                            />
                            <Field
                              name={`activities.${index}.description`}
                              label="Description"
                              component={TextField}
                              className={classes.field}
                            />
                            <div className={classes.verticalCenter}>
                              <IconButton
                                type="button"
                                id={`remove-activity-${index}`}
                                color="secondary"
                                className={classes.iconButton}
                                onClick={() => arrayHelpers.remove(index)}
                              >
                                <Close />
                              </IconButton>
                            </div>
                          </div>
                        </Grid>
                      );
                    })                 
                  }
                  <Grid item xs={12} className={cx(classes.row,classes.footerRow)}>
                    <div className={classes.lunchBox}>
                      <Field
                        name='lunch'
                        label1="Lunch"
                        label2=" "
                        fullWidth
                        margin="none"
                        component={Time}
                        className={classes.field}
                      />
                    </div>
                    <Button
                      color="primary"
                      variant="contained"
                      id={'add-activity'}
                      onClick={() =>
                        arrayHelpers.push({
                          projectId: Object.keys(projects)[0],
                          projectTaskId: -1,
                          length: 0,
                          description: ''
                        })
                      } 
                    >
                      Add Activity
                    </Button>
                  </Grid>
                </Grid>
              );
            }
            }
          />

          <Grid item xs={12} className={cx(classes.row,classes.footerRow)}>
            <Typography variant="h5" margin="none">
              Time Left: {minutesToString(timeLeft)}
            </Typography>
            <Typography
              color="error"
              variant="button"
              className={classes.field}
            >
              {generalError}
            </Typography>
            <div>
              <Button
                type="submit"
                color="primary"
                disabled={isSubmitting || Object.keys(errors).length !== 0 || timeLeft !== 0}
                variant="contained"
                className={classes.button}
              >
                Save
              </Button>

              <Button
                onClick={() => {
                  resetForm(initialValues);
                }}
                id={'shift-edit-reset-button'}
                color="secondary"
                variant="text"
                className={classes.button}
              >
                Reset
              </Button>
            </div>
          </Grid>

        </Grid>
      </Form>
    );
  }
}
ShiftEdit.propTypes = {
  classes: PropTypes.object.isRequired,
  type: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  removeShift: PropTypes.func,
  resetForm: PropTypes.func.isRequired,
  initialValues: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired,
  projects: PropTypes.array.isRequired, 
  projectTasks: PropTypes.array.isRequired, 
  employees: PropTypes.array.isRequired,
  timeLeft: PropTypes.number.isRequired,
  generalError: PropTypes.string,
  values: PropTypes.object.isRequired
};

export default withStyles(styles)(ShiftEdit);