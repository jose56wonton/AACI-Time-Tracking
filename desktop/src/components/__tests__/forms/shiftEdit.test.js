import React from 'react';
import { shallow } from 'enzyme';
import { FieldArray } from 'formik';

import { ShiftEdit } from 'components/forms/ShiftEdit/shiftEdit';
import ShiftEditHOC from 'components/forms/ShiftEdit';

const props =  {  
  classes: {},
  label: 'label',
  type: 'edit',
  isSubmitting: true,
  resetForm: jest.fn(),
  initialValues: {},
  values: {
    activities: [{ id: 0, categoryId: 0, subcategoryId: 0,taskId:0 },{ id: 1, categoryId: 1, subcategoryId: 1,taskId:1 }]
  },
  projects: [{ id: 0 },{ id: 1 },{ id: 2 }],
  projectTasks: [{ id: 0, task: { id: 0,category: { id: 0,name: 'name0' }, subcategory: { id: 0,name: 'name0' } } },
    { id: 1, task: { id: 1,category: { id: 1,name: 'name1' }, subcategory: { id: 1,name: 'name1' } } },
    { id: 2, task: { id: 2,category: { id: 2,name: 'name2' }, subcategory: { id: 2,name: 'name2' } }  }],
  employees: [{ id: 0 },{ id: 1 },{ id: 2 }],
  timeLeft: 60,
  generalError: 'General Error',
  errors: {
    asdf:{
      
    }
  }
};

const renderProps = { 
  remove: jest.fn(),
  push: jest.fn(),
  form: {
    setFieldValue: jest.fn()
  }
};


const setup = overRides => {  
  return shallow(<ShiftEdit {...props} {...overRides}/>);    
};

const setupHOC = overRides => {
  return shallow(<ShiftEditHOC {...props} {...overRides}/>);
};

const setupWithRender = overRides => {
  const wrapper = setup();  
  const Render = wrapper.find(FieldArray).first().prop('render');  
  return shallow(<Render {...renderProps} {...overRides}/>);
};

describe('ShiftEdit Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should render correctly', () => {
    setup();        
  });
  it('should render correctly withStyles', () => {
    setupHOC();    
  });
  it('should call render of activites fieldarray', () => {
    setupWithRender();
  });
  it('should call onChange project Field', () => {
    const wrapper = setupWithRender();
    expect(renderProps.form.setFieldValue).toHaveBeenCalledTimes(0);
    wrapper.find('#project-field-1').first().simulate('change');
    expect(renderProps.form.setFieldValue).toHaveBeenCalledTimes(1);
    expect(renderProps.form.setFieldValue).toHaveBeenLastCalledWith('activities.1.projectTaskId',-1);
  });  
  it('should remove an activity when a remove-projectTask-1', () => {
    const wrapper = setupWithRender();
    expect(renderProps.remove).toHaveBeenCalledTimes(0);
    wrapper.find('#remove-activity-1').first().simulate('click');
    expect(renderProps.remove).toHaveBeenCalledTimes(1);
    expect(renderProps.remove).toHaveBeenCalledWith(1);
  });
  it('should add a activity when add-activity is clicked', () => {
    const wrapper = setupWithRender();
    expect(renderProps.push).toHaveBeenCalledTimes(0);
    wrapper.find('#add-activity').first().simulate('click');
    expect(renderProps.push).toHaveBeenCalledTimes(1);
    expect(renderProps.push).toHaveBeenCalledWith({
      projectId: '0',
      projectTaskId: -1,
      length: 0,
      description: ''
    });
  });
  it('should call resetForm on Reset button click', () => {
    const wrapper = setup();
    expect(props.resetForm).toHaveBeenCalledTimes(0);
    wrapper.find('#shift-edit-reset-button').first().simulate('click');
    expect(props.resetForm).toHaveBeenCalledTimes(1);
    expect(props.resetForm).toHaveBeenCalledWith({});
  });
});