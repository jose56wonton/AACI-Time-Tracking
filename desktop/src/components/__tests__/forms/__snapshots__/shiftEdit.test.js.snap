// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShiftEdit Component should render correctly 1`] = `
<Form>
  <WithStyles(Grid)
    container={true}
    spacing={24}
  >
    <WithStyles(Grid)
      className=""
      item={true}
      xs={12}
    >
      <WithStyles(Typography)
        variant="h6"
      >
        label
      </WithStyles(Typography)>
      <WithStyles(Tooltip)
        title="Delete"
      >
        <WithStyles(IconButton)
          aria-label="Delete"
        >
          <pure(DeleteIcon) />
        </WithStyles(IconButton)>
      </WithStyles(Tooltip)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <C
        component={[Function]}
        fullWidth={true}
        items={
          Array [
            Object {
              "id": 0,
            },
            Object {
              "id": 1,
            },
            Object {
              "id": 2,
            },
          ]
        }
        label="Employee"
        name="employeeId"
      />
      <C
        component={[Function]}
        helper="normal"
        label="Start Date"
        margin="none"
        name="clockInDate"
        type="datetime-local"
      />
      <C
        component={[Function]}
        defaultValue="2017-05-24T10:30"
        helper="normal"
        label="Start Date"
        margin="none"
        name="clockOutDate"
        type="datetime-local"
      />
    </WithStyles(Grid)>
    <C
      name="activities"
      render={[Function]}
    />
    <WithStyles(Grid)
      className=""
      item={true}
      xs={12}
    >
      <WithStyles(Typography)
        margin="none"
        variant="h5"
      >
        Time Left: 
        1:00
      </WithStyles(Typography)>
      <WithStyles(Typography)
        color="error"
        variant="button"
      >
        General Error
      </WithStyles(Typography)>
      <div>
        <WithStyles(Button)
          color="primary"
          disabled={true}
          type="submit"
          variant="contained"
        >
          Save
        </WithStyles(Button)>
        <WithStyles(Button)
          color="secondary"
          id="shift-edit-reset-button"
          onClick={[Function]}
          variant="text"
        >
          Reset
        </WithStyles(Button)>
      </div>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</Form>
`;

exports[`ShiftEdit Component should render correctly withStyles 1`] = `
<ShiftEdit
  classes={
    Object {
      "body": "ShiftEdit-body-6",
      "bodyRow": "ShiftEdit-bodyRow-4",
      "button": "ShiftEdit-button-8",
      "field": "ShiftEdit-field-7",
      "footerRow": "ShiftEdit-footerRow-5",
      "gridContainer": "ShiftEdit-gridContainer-1",
      "headerRow": "ShiftEdit-headerRow-3",
      "lunchBox": "ShiftEdit-lunchBox-10",
      "row": "ShiftEdit-row-2",
      "verticalCenter": "ShiftEdit-verticalCenter-9",
    }
  }
  employees={
    Array [
      Object {
        "id": 0,
      },
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
    ]
  }
  errors={
    Object {
      "asdf": Object {},
    }
  }
  generalError="General Error"
  initialValues={Object {}}
  isSubmitting={true}
  label="label"
  projectTasks={
    Array [
      Object {
        "id": 0,
        "task": Object {
          "category": Object {
            "id": 0,
            "name": "name0",
          },
          "id": 0,
          "subcategory": Object {
            "id": 0,
            "name": "name0",
          },
        },
      },
      Object {
        "id": 1,
        "task": Object {
          "category": Object {
            "id": 1,
            "name": "name1",
          },
          "id": 1,
          "subcategory": Object {
            "id": 1,
            "name": "name1",
          },
        },
      },
      Object {
        "id": 2,
        "task": Object {
          "category": Object {
            "id": 2,
            "name": "name2",
          },
          "id": 2,
          "subcategory": Object {
            "id": 2,
            "name": "name2",
          },
        },
      },
    ]
  }
  projects={
    Array [
      Object {
        "id": 0,
      },
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
    ]
  }
  resetForm={[Function]}
  timeLeft={60}
  type="edit"
  values={
    Object {
      "activities": Array [
        Object {
          "categoryId": 0,
          "id": 0,
          "subcategoryId": 0,
          "taskId": 0,
        },
        Object {
          "categoryId": 1,
          "id": 1,
          "subcategoryId": 1,
          "taskId": 1,
        },
      ],
    }
  }
/>
`;
